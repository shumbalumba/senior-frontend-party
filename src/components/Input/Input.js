// @flow

import * as React from 'react';
import styled from 'styled-components';

import { breakpoints } from '../../theme/media';

const InputWrapper = styled.div`
  position: relative;
`;

const InputStyled = styled.input`
  margin: 10px 0;
  width: ${props => props.theme.width.formElementRegular};
  height: ${props => props.theme.height.regular};
  padding-left: 50px;
  color: ${props => props.theme.colour.grey};

  @media (max-width: ${breakpoints.sm}) {
    width: ${props => props.theme.width.formElementSmall};
  }

  ::placeholder {
    color: ${props => props.theme.colour.grey};
  }

  &:focus {
    color: ${props => props.theme.colour.grey};
  }
`;

const IconWrapper = styled.div`
  svg {
    fill: ${props => props.theme.colour.lightGrey};
    height: 16px;
    position: absolute;
    top: 50%;
    left: 25px;
    transform: translateY(-50%);
  }
`;

type InputProps = {
  type: string,
  placeholder?: string,
  icon?: ?React.Node,
};

export const Input = ({ icon, ...rest }: InputProps) => (
  <InputWrapper>
    {icon && <IconWrapper>{icon}</IconWrapper>}
    <InputStyled {...rest} className="form-control" />
  </InputWrapper>
);

Input.defaultProps = {
  placeholder: '',
  icon: null,
};
