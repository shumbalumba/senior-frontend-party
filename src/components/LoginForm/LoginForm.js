// @flow

import * as React from 'react';
import styled from 'styled-components';

import { Input } from '../Input/Input';
import { Button } from '../Button/Button';
import { UserIcon, PasswordIcon } from '../Icons/Icons';
import type { LoginActionPayload } from '../../store/actions/LoginAction';

const Form = styled.form`
  margin-top: 50px;
  display: grid;
  justify-items: center;
`;

type LoginFormProps = {
  onSubmit?: (payload: LoginActionPayload) => void,
};

type LoginFormState = {
  username: string,
  password: string,
};

export class LoginForm extends React.Component<LoginFormProps, LoginFormState> {
  static defaultProps = {
    onSubmit: null,
  };

  state = {
    username: '',
    password: '',
  };

  handleChange = (e: SyntheticInputEvent<HTMLInputElement>) => {
    const { name, value } = e.currentTarget;
    this.setState(state => ({ ...state, [name]: value }));
  };

  handleSubmit = (e: SyntheticEvent<HTMLFormElement>) => {
    const { onSubmit } = this.props;
    const { username, password } = this.state;
    if (onSubmit) onSubmit({ username, password });
    e.preventDefault();
  };

  render() {
    const { username, password } = this.state;
    return (
      <Form onSubmit={this.handleSubmit}>
        <Input
          type="text"
          placeholder="Username"
          name="username"
          icon={<UserIcon />}
          onChange={this.handleChange}
          value={username}
        />
        <Input
          type="password"
          placeholder="Password"
          name="password"
          icon={<PasswordIcon />}
          onChange={this.handleChange}
          value={password}
        />
        <Button text="Log In" type="submit" />
      </Form>
    );
  }
}
