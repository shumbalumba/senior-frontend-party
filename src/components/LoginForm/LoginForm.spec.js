import * as React from 'react';
import { shallow } from 'enzyme';
import { LoginForm } from './LoginForm';
import { Input } from '../Input/Input';

describe('<LoginForm />', () => {
  const onSubmit = jest.fn();
  const event = {
    preventDefault: jest.fn(),
  };

  it('renders', () => {
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);
    expect(wrapper.exists()).toBeTruthy();
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders two inputs', () => {
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);
    expect(wrapper.find(Input)).toHaveLength(2);
  });

  it('calls onSubmit', () => {
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);
    const Form = wrapper.dive().find('form');
    Form.simulate('submit', event);
    expect(onSubmit).toBeCalled();
  });

  it('handles state change', () => {
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);
    const UsernameInputEvent = {
      currentTarget: {
        name: 'username',
        value: 'username',
      },
      preventDefault: jest.fn(),
    };
    wrapper.instance().handleChange(UsernameInputEvent);
    expect(wrapper.state().username).toEqual('username');
  });

  it('calls submit with correct values', () => {
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);
    const stateValues = {
      username: 'username',
      password: 'password',
    };
    wrapper.setState(stateValues);
    const Form = wrapper.dive().find('form');
    Form.simulate('submit', event);
    expect(onSubmit).toBeCalledWith(stateValues);
  });
});
