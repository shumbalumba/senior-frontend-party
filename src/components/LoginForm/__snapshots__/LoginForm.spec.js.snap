// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> matches snapshot 1`] = `
<styled.form
  onSubmit={[Function]}
>
  <Input
    icon={<Unknown />}
    name="username"
    onChange={[Function]}
    placeholder="Username"
    type="text"
    value=""
  />
  <Input
    icon={<Unknown />}
    name="password"
    onChange={[Function]}
    placeholder="Password"
    type="password"
    value=""
  />
  <Component
    text="Log In"
    type="submit"
  />
</styled.form>
`;
